### PR TITLE
Ignore SendMessageTimeout failures on Windows Nano Server

### LINKED ISSUE #30566: EnvironmentVariable operations with EnvironmentVariableTarget.User fails on Windows Nano Server
This seems to be a product bug because of an underlying windows api change.

cc @Anipik @stephentoub @danmosemsft

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #30566

### CODE REVIEW

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.GetEnvironmentVariable.cs [hunk 1]
```diff
@@ -220,12 +220,6 @@ public void EnumerateEnvironmentVariables(EnvironmentVariableTarget target)
             bool lookForSetValue = (target == EnvironmentVariableTarget.Process)
                 || (PlatformDetection.IsWindows && PlatformDetection.IsPrivilegedProcess);
 
-            // [ActiveIssue("https://github.com/dotnet/runtime/issues/30566")]
-            if (PlatformDetection.IsWindowsNanoServer && target == EnvironmentVariableTarget.User)
```
Comment (danmoseley) [inline src/libraries/System.Runtime.Extensions/tests/System/Environment.GetEnvironmentVariable.cs:224]: I guess this has to stay in SetEnvironmentVariable?

https://github.com/dotnet/runtime/blob/555369171d198c49d45051e61dee2797ab0561fb/src/libraries/System.Runtime.Extensions/tests/System/Environment.SetEnvironmentVariable.cs#L19

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Environment.Win32.cs [hunk 2]
```diff
@@ -56,7 +56,9 @@ private static void SetEnvironmentVariableFromRegistry(string variable, string?
                 {
                     IntPtr unused;
                     IntPtr r = Interop.User32.SendMessageTimeout(new IntPtr(Interop.User32.HWND_BROADCAST), Interop.User32.WM_SETTINGCHANGE, IntPtr.Zero, (IntPtr)lParam, 0, 1000, &unused);
-                    Debug.Assert(r != IntPtr.Zero, $"SetEnvironmentVariable failed: {Marshal.GetLastPInvokeError()}");
+
+                    // SendMessageTimeout message is a empty stub on Windows Nano Server that fails with both result and last error 0.
+                    Debug.Assert(r != IntPtr.Zero || Marshal.GetLastPInvokeError() == 0, $"SetEnvironmentVariable failed: {Marshal.GetLastPInvokeError()}");
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Environment.Win32.cs:61]: the docs say "If target is [User](https://learn.microsoft.com/en-us/dotnet/api/system.environmentvariabletarget?view=net-7.0#system-environmentvariabletarget-user) or [Machine](https://learn.microsoft.com/en-us/dotnet/api/system.environmentvariabletarget?view=net-7.0#system-environmentvariabletarget-machine), other applications are notified of the set operation by a Windows WM_SETTINGCHANGE message."

I don't know whether it's worth saying "except for Nano"..

File: src/libraries/System.Runtime.Extensions/tests/System/Environment.GetEnvironmentVariable.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Runtime.Extensions/tests/System/Environment.GetEnvironmentVariable.cs:224]: actually I think that should be removed 

https://github.com/dotnet/runtime/blob/555369171d198c49d45051e61dee2797ab0561fb/src/libraries/System.Runtime.Extensions/tests/System/Environment.SetEnvironmentVariable.cs#L19-L23

File: src/libraries/System.Private.CoreLib/src/System/Environment.Win32.cs (hunk 2 above)
Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Environment.Win32.cs:61]: Nothing that uses HWND messages works on Nano Server (including e.g. `ApartmentState.STA`). This is just a corner case.

There are so many things that do not work on Nano Server. Documenting them would take some effort. We should wait for somebody asking for it first.

**Pull request merged**